Impelement at_least, at_most, times, with, verify reset methods

ka2u opened the pull request:
I added some tests and fixed the with method.
Thank you for your time and attention.

ka2u committed:
Impelement at_least, at_most, times, with, verify reset methods

masaki commented:
気になったところを。日本語で失礼しますよ。。。
- with の確認のところ、List::MoreUtils の each_array を使ってもいいのでは。
- verify, reset を EXPORT するのは、名前が汎用すぎるのであまり良くない気がします。
- 02_mock.t で use Test::Exception してるけど使ってない気がします。
- 02_mock.t に at_least とか全部つめ込まないで、02_mock/at_least.t とか分けちゃってもいいと思いました。
- verify_fail のテストは、それこそ Test::Exception の dies_ok や throws_ok で書けないですかね。

ka2u commented:
ありがとうございます。
下記の様に変更しようと思います。また2番目と５番目はご意見を伺いたいのですがいかがでしょうか。
- List::MoreUtilsのeach_arrayに変更します
- verify、resetをクラス・メソッドにしようかと思いますがどうでしょうか？  ( Test::Double->verify, Test::Double->reset ) mockingしたオブジェクトには同名のメソッドがあった場合の問題があるので避けました。
- 02_mock.tのTest::Exceptionを削除します
- テストを分けるようにします
- verifyしたときに意図した呼び出しでない場合はdieしたほうがいいでしょうか？ Testという名前空間なのでTest::Builder(Test::More)の結果を使ったほうがいいかと思っています

ka2u commented:
すいません、いまさらなんですが最初のpull-requestでMochaって書こうとしたのにMockitoと書いてしまいました。。

masaki commented:
5番目について
verify() は戻り値が bool のメソッドにしておいて、verify_ok() は verify() を呼んで ok/not ok する、というのはどうでしょうか。
意図しない場合に die するというのは、暗に RSpec とかのイメージを持込み過ぎた感があるので、忘れて下さい。

2番目はクラスメソッドでいいんじゃないかと思います。ただし上の話と被ってきますね。

ka2u commented:
良いと思います。では
Test::Double->verify
Test::Double->verify_ok
Test::Double->reset
を用意しようと思います。

ka2u committed:
Move export methods to class methods and Add verify_result method

verify and reset methods move to class methods.
Add returns verify result method.
Tests were divided.

ka2u commented:
間が空いてしまってすみません。。実装してみましたが、verify結果を返す方をverify_resultメソッドに、ok/not okする方をverifyメソッドにしたいと思いました。理由としましては、verify結果として返す値がハッシュになっていまして、値が多少複雑なため結果を返すほうが使用頻度が低いのではないかと思いました。また、上記の理由から書きやすいverifyの方をok/not okするようにしたいのですがいかかでしょうか。
上記含めた修正をpushしましたので、お忙しいところすみませんが、レビューしていただけますと助かります。

Test::Double->verify # on/not okまで行う
Test::Double->verify_result # verifyした結果のハッシュが返る

masaki commented:
@ka2u Test::Double の verify の出力を capture してるテストするところですが、Test::Output や Capture::Tiny あたりでできたりしないですかね。

ka2u commented:
ちょっと検討してみます。確かに今のはちょっとわかりにくい方法だと思います。ありがとうございます

masaki commented:
テストコードにおさまりがつけば、あとは問題ないと思います。実装面は特に。

ka2u committed:
Use Test::Tester for TAP output test

ka2u commented:
存外Test::Output、Capture::Tinyでうまく出力がとれず、代わりにTest::Testerでテストするようにしました。
内容はわかりやすく、書きやすくなっていると思います。ちなみにTest::TesterはTest::Outputのテストに使われていました。

masaki commented on t/lib/Test/Simple/Catch.pm:
```diff
@@ -0,0 +1,20 @@
+# For testing Test::Simple;
```
Test::Tester に変更したから多分不要ですよねこのファイル。

masaki commented on t/lib/TieOut.pm:
```diff
@@ -0,0 +1,30 @@
+package TieOut;
```
これも同じく消していいかな？

masaki commented:
@ka2u Test::Tester に置き換えて不要になったファイル2つ？を消したらマージしましょう！

ka2u committed:
Delete unnecessary test file

ka2u committed:
Fix Readme

ka2u commented:
ありがとうございます。消すの忘れておりました。ちょっとReadmeも直しました。もう、マージしてしまってよろしいですか？
